chapters/interactions-nonlinear.qmd

```diff
@@ -789,6 +789,196 @@ The H-statistic and its alternatives can be an incredibly valuable tool for lear
+    mod <- lm(y ~ 1, data = tr)
+  } else {
+    mod <- lm(y ~ poly(x, deg), data = tr)
+  }
+  te$.pred <- predict(mod, newdata = te)
+  te$degree <- deg
+  te
+}
+sim <- function(deg, n = 30) {
+  make_data(num_points = n) %>% 
+    get_preds(deg)
+}
+```
+
+To create a simulated data set, `r nrow(make_data())` samples were simulated. Of these, `r nrow(make_data()) - 1` were used for the training set and one were reserved for the demonstrating the variance and bias. The test set values were roughly equally spaced across the range of the predictor. One simulated data set is shown in @fig-nonlinear-sim, along with the true underlying pattern.  
+
+```{r}
+#| label: fig-nonlinear-sim
+#| echo: false
+#| out-width: 70%
+#| fig-width: 6
+#| fig-height: 4.25
+#| fig-cap: "A simulated data set with a nonlinear function."
+sim_grid <- 
+  tibble(x = seq(from = -1.2, to = 1.2, length.out = 500)) %>% 
+  mutate(y = (x^3) + 2 * exp(-6 * (x - 0.3)^2))
+set.seed(12)
+make_data() %>% 
+  ggplot(aes(x, y)) + 
+  geom_line(data = sim_grid, lty = 3) +
+  geom_point(aes(col = data, pch = data), cex = 2, alpha = 3 / 4) +
+  labs(x = "predictor", y = "outcome")
+```
+
+
+```{r}
+#| label: variance-bias-stats
+#| echo: false
+#| cache: true
+set.seed(614)
+sim_res <- 
+  map_dfr(rep(1:23, each = 50000), sim) %>%
+  mutate(bias = truth - .pred) %>% 
+  summarize(
+    rmse = rmse_vec(y, .pred),
+    bias = mean(bias),
+    var = var(.pred),
+    n  = length(.pred),
+    .by = c(degree, point)
+  ) %>% 
+  mutate(
+    double_check = var + bias^2, 
+    mse = rmse^2, 
+    point = format(point)
+  )
+```
+
+A linear model was estimated using ordinary least squared and a polynomial expansion for each simulated data set. The test set values were predicted, and the bias, variance, and root mean squared error statistics were computed. This was repeated 50,000 times for polynomial degrees ranging from one to twenty.  
+
+@fig-sim-results shows the average statistic values across model complexity (i.e., polynomial degree). A linear model performs poorly for the bias due to underfitting (as expected). In panel (a), Adding more nonlinearity results in a substantial decrease in bias because the model fit is closer to the true equation. This improvement plateaus at a sixth-degree polynomial and stays low (nearly zero). The variance begins low; even though the linear model is ineffective, it is stable. Once additional terms are added, the variance of the model steadily increases then explodes around an 20<sup>th</sup> degree polynomial^[In this particular case, the variance becomes very large since the number of parameters is nearly the same as the number of training set points (`r nrow(make_data()) - 1`). This makes the underlying mathematical operation (matrix inversion) numerically unstable. Even so, it is the result of excessive model complexity.]. This shows the tradeoff; as the model becomes more complex, it fits the data better but eventually becomes unstable.
```

Comment:
```suggestion
@fig-sim-results shows the average statistic values across model complexity (i.e., polynomial degree). A linear model performs poorly for the bias due to underfitting (as expected). In panel (a), adding more nonlinearity results in a substantial decrease in bias because the model fit is closer to the true equation. This improvement plateaus at a sixth-degree polynomial and stays low (nearly zero). The variance begins low; even though the linear model is ineffective, it is stable. Once additional terms are added, the variance of the model steadily increases then explodes around an 20<sup>th</sup> degree polynomial^[In this particular case, the variance becomes very large since the number of parameters is nearly the same as the number of training set points (`r nrow(make_data()) - 1`). This makes the underlying mathematical operation (matrix inversion) numerically unstable. Even so, it is the result of excessive model complexity.]. This shows the tradeoff; as the model becomes more complex, it fits the data better but eventually becomes unstable.
```